File: SampSharpGameMode/Story/TimeOfDay.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using SyntheticVideo2language.Story.Api;

namespace SampSharp.SyntheticGameMode.Story
{
    public class TimeOfDay : StoryTimeOfDayBase
    {
        public int Hour { get; set; }
        public int Minutes { get; set; }

        public TimeOfDay(int hour, int minutes)
        {
            this.Hour = hour;
            this.Minutes = minutes;
        }

        public override string Description
        {
            get
            {
                var time = (this.Hour + (this.Minutes < 30 ? 0 : 1)) % 24;
                if (time > 5 && time < 11)
                {
                    return "in the morning";
                }
                else if (time == 12)
                {
                    return "at noon";
                }
                else if (time > 11 && time < 16)
                {
                    return "during the day";
                }
                else if (time > 16 && time < 19)
                {
                    return "in the evening";
                }
                else if (time == 0)
                {
                    return "in the middle of the night";
                }
                else
                {
                    return "during the night";
                }
            }
            set { }
        }

        public async override Task<bool> ApplyAsync(params object[] parameters)
        {
            if (parameters.Length != 1)
            {
                throw new IndexOutOfRangeException();
            }

            var player = parameters[0] as Player;
            await Task.Delay(100);

            player.SetTime(this.Hour, this.Minutes);
            SampStory.Instance.Logger.Log(" " + this.Description, player);
            await Task.Delay(100);

            return true;
        }
    }
}

File: story/TimeOfDay.lua
TimeOfDay = class(StoryTimeOfDayBase, function(o, hour, minute)
    local time = hour % 24
    if minute >= 30 then
        time = (hour + 1) % 24
    end
    local description = ""
    if (time > 5 and time < 11) then
        description = "in the morning"
    elseif (time == 12) then
        description = "at noon";
    elseif (time > 11 and time < 16) then
        description = "during the day"
    elseif (time > 16 and time < 19) then
        description = "in the evening"
    elseif (time == 0) then
        description = "in the middle of the night"
    else
        description = "during the night"
    end

    StoryTimeOfDayBase.init(o, description)
    o.Hour = hour
    o.Minute = minute
end)

function TimeOfDay:Apply(...)
    local player = nil
    for i,v in ipairs(arg) do
        player = v
        break
    end

    setTime(self.Hour, self.Minute)
    local id = player:getData("id")
    local scenarioId = player:getData("scenarioId")
    local story = STORIES[id][scenarioId]

    if not story then
        outputConsole("Error: the story is null "..id.." ---> "..scenarioId)
    end
    story.Logger.Log(" " .. self.Description, player)
end